File: Examples/CSharp/x-IMU3-Examples/Examples/OpenAsync.cs
namespace Ximu3Examples
{
    class OpenAsync
    {
        public OpenAsync()
        {
            // Search for connection
            var devices = Ximu3.PortScanner.ScanFilter(Ximu3.CApi.XIMU3_PortType.XIMU3_PortTypeUsb);

            if (devices.Length == 0)
            {
                Console.WriteLine("No USB connections available");
                return;
            }

            Console.WriteLine("Found " + Ximu3.Helpers.ToString(Ximu3.CApi.XIMU3_device_to_string(devices[0])));

            // Open connection
            var connection = new Ximu3.Connection(Ximu3.ConnectionConfig.From(devices[0])!);

            Ximu3.Connection.OpenAsyncCallback callback = result =>
            {
                if (result != Ximu3.CApi.XIMU3_Result.XIMU3_ResultOk)
                {
                    Console.WriteLine("Unable to open " + connection.GetConfig() + ". " + Ximu3.Helpers.ToString(Ximu3.CApi.XIMU3_result_to_string(result)) + ".");
                    return;
                }

                Console.WriteLine(Ximu3.Helpers.ToString(Ximu3.CApi.XIMU3_result_to_string(result)));
            };

            connection.OpenAsync(callback);

            // Close connection
            System.Threading.Thread.Sleep(3000);

            connection.Close();
        }
    }
}

File: Examples/Python/open_async.py
import time

import ximu3

# Search for connection
devices = ximu3.PortScanner.scan_filter(ximu3.PORT_TYPE_USB)

time.sleep(1)  # wait for OS to release port

if not devices:
    raise Exception("No USB connections available")

print(f"Found {devices[0]}")

# Open connection
connection = ximu3.Connection(devices[0].connection_config)


def callback(result: int) -> None:
    if result != ximu3.RESULT_OK:
        raise Exception(f"Unable to open {connection.get_config()}. {ximu3.result_to_string(result)}.")

    print(ximu3.result_to_string(result))


connection.open_async(callback)

# Close connection
time.sleep(3)

connection.close()
